vnext/Desktop/WinHTTPWebSocketResource.cpp

```diff
@@ -0,0 +1,447 @@
+
+// clang-format off
+#include "pch.h"
+
+#include <Unicode.h>
+#include <Utils.h>
+#include "WinHTTPWebSocketResource.h"
+
+// Windows API
+#include <comutil.h>
+
+using Microsoft::Common::Unicode::Utf8ToUtf16;
+
+using std::function;
+using std::size_t;
+using std::string;
+using std::wstring;
+
+namespace Microsoft::React
+{
+
+WinHTTPWebSocketResource::WinHTTPWebSocketResource(const string& urlString)
+  : m_openSessionFlags{ WINHTTP_FLAG_ASYNC }
+  , m_sessionHandle{ NULL }
+  , m_connectionHandle{ NULL }
+  , m_requestHandle{ NULL }
+  , m_errorStatus{ ERROR_SUCCESS }
+{
+#if 0 // WinHttpCrackUrl does not accept non-HTTP schemes 🙄
+  BSTR bstr = _com_util::ConvertStringToBSTR(urlString.c_str());
+  LPWSTR lpwstr = bstr;
+  SysFreeString(bstr);
+
+  LPURL_COMPONENTS url{};
+  auto parseResult = WinHttpCrackUrl(lpwstr, static_cast<DWORD>(urlString.length()), /*flags*/ 0, url);
+  if (!parseResult)
+    throw std::exception("Could not parse URL.");
+
+  if (wcscmp(L"wss", url->lpszScheme) == 0)
+  {
+    m_openSessionFlags |= WINHTTP_FLAG_SECURE;
+  }
+  else if (wcscmp(L"ws", url->lpszScheme) != 0)
+  {
+    throw std::exception(Common::Unicode::Utf16ToUtf8(
+      std::wstring(L"Incorrect URL scheme: ") + std::wstring(url->lpszScheme)
+    ).c_str());
+  }
+
+  m_url = *url;
+#else
+  Url url{ urlString };
+  ZeroMemory(&m_url, sizeof(m_url));
+  m_url.dwStructSize = sizeof(m_url);
+  m_url.lpszScheme = _wcsdup(wstring(url.scheme.begin(), url.scheme.end()).c_str());
+  m_url.dwSchemeLength = static_cast<DWORD>(url.scheme.length());
+  m_url.lpszHostName = _wcsdup(wstring(url.host.begin(), url.host.end()).c_str());
+  m_url.dwHostNameLength = static_cast<DWORD>(url.host.length());
+  m_url.nPort = static_cast<INTERNET_PORT>(std::atoi(url.port.c_str()));
+  m_url.lpszUrlPath = _wcsdup(wstring(url.path.begin(), url.path.end()).c_str());
+  m_url.dwUrlPathLength = static_cast<DWORD>(url.path.length());
+  m_url.lpszUserName = NULL;
+  m_url.dwUserNameLength = 0;
+  m_url.lpszPassword = NULL;
+  m_url.dwPasswordLength = 0;
+  m_url.lpszExtraInfo = NULL;
+  m_url.dwExtraInfoLength = 0;
+#endif // 0
+}
+
+WinHTTPWebSocketResource::~WinHTTPWebSocketResource() /*override*/
+{
+  if (m_webSocketHandle != NULL)
+  {
+    WinHttpCloseHandle(m_webSocketHandle);
+    m_webSocketHandle = NULL;
+  }
+
+  if (m_requestHandle != NULL)
+  {
+    WinHttpCloseHandle(m_requestHandle);
+    m_requestHandle = NULL;
+  }
+
+  if (m_connectionHandle != NULL)
+  {
+    WinHttpCloseHandle(m_connectionHandle);
+    m_connectionHandle = NULL;
+  }
+
+  if (m_sessionHandle != NULL)
+  {
+    WinHttpCloseHandle(m_sessionHandle);
+    m_sessionHandle = NULL;
+  }
+}
+
+#pragma region IWebSocketResource
+
+void WinHTTPWebSocketResource::Connect(const Protocols& protocols, const Options& options)
+{
+  m_sessionHandle = WinHttpOpen(
+    L"React Native Windows",              // pszAgentW
+    WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY,  // dwAccessType
```

Comment:
Is this advisable? Most websites do different things for different agent strings (e.g. render different pages for mobile vs. desktop). I'd rather either:
1) Go with a safe default like Chromium's agent string
2) Go with a safe default but provide a way to override this by the app developer